Output result as data

petergoes opened the pull request:
The module outputs an object to consuming modules so own reporters can be written. The CLI version should als use this data instead of the `parser-stats` module. However, to keep this PR small. That is a done in its own PR (not started as of the time of this writing)

To test this PR, replace the current `.then` in `bin/fastatic` with:

```
.then(output => {
    console.log('src:')
    Object.keys(output.fileSize.src).map(key => {
        console.log(' ', key, filesize(output.fileSize.src[key]))
    })
    console.log('dest:')
    Object.keys(output.fileSize.dest).map(key => {
        console.log(' ', key, filesize(output.fileSize.dest[key]))
    })
    console.log('output:');
    console.log(output);
})
```

closes #70 (and supersedes #75)

petergoes committed:
refactor(parse-all): report error per parser

When a parser errors, report what happend to the user. Continue with all
parsers before exiting. Parsers are started in paralell, we need to wait
for all of them to complete to clean up the temp folder.

petergoes committed:
refactor(index): return the promise chain

By returning the promise chain consuming modules can act on the
resolving or rejecting of the promise chain. As a result, the
bin/fastatic module can set the exitCode as discussed in #75

petergoes committed:
Merge branch 'master' into fix/exit-process-after-error

petergoes committed:
Merge branch 'master' into fix/exit-process-after-error

petergoes committed:
refactor(index): split execution from cleanup flow

To be able to return the result to consuming modules, the result of the
parsing should be returned before the cleanup is executed

petergoes committed:
feat(results): add results module

The results module returns an object with stats. The fastatic module
returns this object to consuming modules

petergoes committed:
refactor(results): split functions in smaller pieces

petergoes committed:
Merge branch 'master' into output-result-as-data

petergoes committed:
refactor(results): remove saved bytes from result

petergoes committed:
refactor(results): rename to compareFileSize

petergoes committed:
refactor(compare-file-size): accept two dirs instead of config object

petergoes committed:
refactor(compare-file-size): return sizes object per folder

petergoes committed:
refactor(compare-file-size): put calculating file sizes in own module

petergoes commented:
@jbmoelker I split the module up and return a new object as discussed

petergoes committed:
refactor(index): return an object with the fileSizes as a property

Now we can output more data to consuming modules

jbmoelker commented on index.js:
```diff
@@ -14,17 +16,26 @@ function fastatic(options) {
+		.then(() => stats(config))
+		.then(output => console.log(output))
+		.then(() => loader.stop())
+		.then(() => Promise.all([
```
I agree this would be the setup if we do more than `compareFileSize`, but without it, I find this overcomplicating the module. So leave out `spread` and instead use:

``` javascript
.then(() => ({
  fileSize: compareFileSize(config.src, config.temp)
});
```

jbmoelker commented on lib/compare-file-size.js:
```diff
@@ -0,0 +1,24 @@
+function compareFileSize(sourceFolder, tempFolder) {
+	return glob('**/*', { nodir: true, cwd: sourceFolder })
+		.then(files => {
+			const sourceFiles = appendDirToFiles(sourceFolder, files);
```
I'd use `const sourceFiles = files.map(file => path.join(sourceFolder, file));` The function `appendDirToFiles` doesn't instantly tell me what will happen inside. So for a oneline function I find it too much.

jbmoelker commented on lib/compare-file-size.js:
```diff
@@ -0,0 +1,24 @@
+const glob = promisify(require('glob').glob);
+const getFileSizesPerType = require('./get-filesizes-per-type');
+
+function compareFileSize(sourceFolder, tempFolder) {
```
this module doesn't need to know about `temp`, and to stick with other vocabulary I would use `(srcDir, destDir)`.

And later also `srcFiles` and `destFiles`. That's why all the way in the end you also return `{ src: ..., dest: ... }`.

petergoes replied:
Although I do agree that term `tempFolder` is wrong, I do think this module needs to know about the two folders. We use it here to compare the file sizes of the source and dest folders, but this module can be used on any two folders. If we were to take out this module and release it on its own, the naming `sourceFolder` and `tempFolder`/`destFolder` are wrong, but I don't know a better naming. Thoughts?

jbmoelker commented on lib/get-filesizes-per-type.js:
```diff
@@ -0,0 +1,25 @@
+function getFileSizesPerType(files) {
+	const sizes = getFileSizes(files);
+
+	return files.reduce((typeSizes, file) => {
```
within this scope you can simplify / shorten your variable names `sizes`, `type`, `size`, `total`. So: `sizes[type] = total + size;` instead of `typeSizes[fileType] = totalForType + fileSize;`.

jbmoelker commented on lib/get-filesizes-per-type.js:
```diff
@@ -0,0 +1,25 @@
+	return files.reduce((typeSizes, file) => {
+		const fileType = path.extname(file).replace('.', '');
+		const fileSize = sizes[file];
+		const totalForType = typeSizes[fileType] || 0;
```
Rather `sizes[type] = (sizes[type] || 0) + size;` ?

jbmoelker commented on lib/parse-all.js:
```diff
@@ -1,6 +1,7 @@
 'use strict';
-
+const colors = require('colors');
```
never used

jbmoelker commented on lib/parse-all.js:
```diff
@@ -1,6 +1,7 @@
-
+const colors = require('colors');
 const Deferred = require('./deferred');
+const path = require('path');
```
never used.

jbmoelker commented on lib/parse-all.js:
```diff
@@ -22,7 +23,10 @@ function parseAll(config) {
-				.catch(err => parsing[name].reject(err));
+				.catch(err => {
+					parsing[name].reject(err);
+					return { name: name, error: err };
```
why do we need to do this?

petergoes replied:
We need to run all the parsers before we can generate error reporting. this way, we store which parser fails (`name`) and why (`error`)

petergoes committed:
fix(index): replace console log with the logger module

petergoes committed:
refactor: create temp folders for src and dist files

We copy the src users source folder to a temp dir. We do that so we can
compare the files in a time when the src folder might be overridden by
the parsed files

petergoes committed:
refactor(index): remove unneeded Promise.all

petergoes committed:
refactor(compare-file-sze): use path module to append dir to files

petergoes committed:
refactor(get-filesizes-per-type): simplify variable names

petergoes committed:
refactor(get-filesizes-per-type): simplify size addition

petergoes committed:
refactor(parse-all): remove unused dependencies

petergoes committed:
fix(get-filesizes-per-type): fix variable naming collisions

jbmoelker commented on bin/fastatic:
```diff
@@ -11,8 +12,15 @@ const program = require('commander')
+})
+.then(config => logger.log('Done. Optimised static files in', config.dest))
+	.catch(err => {
+		console.error(err.message);
```
`console.error` => `logger.error`?

jbmoelker commented on lib/compare-file-size.js:
```diff
@@ -0,0 +1,20 @@
+const glob = promisify(require('glob').glob);
+const path = require('path');
+
+function compareFileSize(sourceFolder, tempFolder) {
```
I'd still refactor to `srcDir`, `destDir`, `srcFiles` and `destFiles` to be consistent. We've never used "source" or "folder" anywhere.

jbmoelker commented on index.js:
```diff
@@ -9,22 +10,27 @@ const stats = require('./lib/parser-stats');
 
-	const logger = new Logger(config.logLevel);
+	result
+		.then(() => remove(config.temp.root));
```
doesn't need a newline

jbmoelker commented on lib/parse-all.js:
```diff
@@ -1,5 +1,4 @@
-
 const Deferred = require('./deferred');
 const parserProgressBar = require('./parser-progress-bar');
 
```
This PR makes me realise that I think we should refactor the signature of parseAll to `parseAll(cwd, parsers)` and no longer expose all the other config settings to this method. But that's for a separate PR :)

jbmoelker commented on index.js:
```diff
@@ -1,3 +1,4 @@
+const compareFileSize = require('./lib/compare-file-size');
```
missing `)`

petergoes committed:
chore: remove unneeded white space

petergoes committed:
refactor(compare-file-size): keep variable names consistant

jbmoelker commented on index.js:
```diff
@@ -9,22 +10,27 @@ const stats = require('./lib/parser-stats');
-	const loader = new Spinner(`%s Crunching ${config.src}`);
-	loader.setSpinnerString(18);
-	loader.start();
+	const result = Promise.all([
```
sorry for the nitpicking, but I think line 16-18 should indent one tab less

jbmoelker commented on bin/fastatic:
```diff
@@ -11,8 +12,15 @@ const program = require('commander')
 
+const logger = new Logger(program.logLevel);
+
 fastatic({
```
more nitpicking: indent line 18-22 with an extra tab?

petergoes committed:
style: add or remove extra tab
